d2ir/compile.go

```diff
@@ -22,12 +22,50 @@ func Compile(ast *d2ast.Map) (*Map, error) {
 	m.initRoot()
 	m.parent.(*Field).References[0].Context.Scope = ast
 	c.compileMap(m, ast)
+	c.compileClasses(m)
 	if !c.err.Empty() {
 		return nil, c.err
 	}
 	return m, nil
 }
 
+func (c *compiler) compileClasses(m *Map) {
+	classes := m.GetField("classes")
+	if classes == nil || classes.Map() == nil {
+		return
+	}
+
+	layersField := m.GetField("layers")
+	if layersField == nil {
+		return
+	}
+	layers := layersField.Map()
+	if layers == nil {
+		return
+	}
+
+	for _, lf := range layers.Fields {
+		if lf.Map() == nil || lf.Primary() != nil {
+			c.errorf(lf.References[0].Context.Key, "invalid layer")
+			continue
+		}
+		l := lf.Map()
+		lClasses := l.GetField("classes")
+
+		if lClasses == nil {
+			lClasses = classes.Copy(l).(*Field)
+			l.Fields = append(l.Fields, lClasses)
```

Comment:
Isn't this too late? By this time we've already interpreted all the classes in the layer?

Reply:
oh hm, i guess a test like this is what u mean
```
classes: {
  orb: {
    style.fill: yellow
  }
}
layers: {
  x: {
    classes.orb.style.stroke: red
  }
}
```

Reply:
Nah I mean this:

```
classes: {
  orb: {
    style.fill: yellow
  }
}
layers: {
  x: {
    class: orb
  }
}
```

Wouldn't `orb` be undefined in `layers.x` as we don't add it until after compilation?

Reply:
no, d2ir's compile actually doesn't look at the `class` keyword at all. the interpretation happens later

Reply:
Oh right!

Reply:
(the example i gave of what i thought u meant remains an issue though)

Reply:
Why? Don't you overlay if classes already exists just below that code?

Reply:
ah fuck, was just a typo in the test case i had. https://github.com/terrastruct/d2/pull/772/commits/352a6a36d6043deefebf54b3b38c704030150e32

you're right. 

d2ir is 👨‍🍳